SQL Auditing: Remove AUDIT_CHANGE_GROUP from available AuditActionGroups

ranisha2 opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
<!--
Please add an informative description that covers the changes made by the pull request.

If applicable, reference the bug/issue that this pull request fixes here.
-->

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

markcowl commented on src/ResourceManager/Sql/Commands.Sql/Auditing/Model/BaseBlobAuditingPolicyModel.cs:
```diff
@@ -18,8 +18,7 @@ public enum AuditActionGroups
             BATCH_COMPLETED_GROUP,
-            APPLICATION_ROLE_CHANGE_PASSWORD_GROUP, 
-            AUDIT_CHANGE_GROUP,
+            APPLICATION_ROLE_CHANGE_PASSWORD_GROUP,
```
So, this is a breaking change, right?  You will need to deprecate these values, and, if provided by the user, map them into other values.  Otherwise, users with scripts that already use these values will be broken.

ranisha2 replied:
@markcowl AUDIT_CHANGE_GROUP was never supported. This action group was mistakenly added, and when provided the backend throws an exception anyway.

ranisha2 replied:
Also, it cannot be mapped to anything, it should just be removed, as it shouldn't have been there in the first place.

cormacpayne replied:
@ranisha2 just to be 100% sure (☺️), if a user provided the `AUDIT_CHANGE_GROUP` value for any parameter that accepts an `AuditActionGroup` enum, then it would have _never_ worked and _always_ thrown an exception?

ranisha2 replied:
@cormacpayne When enabling a policy with AUDIT_CHANGE_GROUP value, our backend will always throw an exception. The only way to provide this value with no exception to be thrown is when disabling the policy, which does not make any sense as this value does not take effect when the policy is disabled.

markcowl replied:
@ranisha2 But could there be a customer script that used this value when disabling a policy?  As long as there can be no working script that uses this value, then this change is OK, otherwise, it is a breaking change, and the proper thing is to deprecate the value rather than removing it.  If possible, for enable scenarios, the value would be mapped to anohter (default) value, although throwing would be OK, if that is what the service does now.

cormacpayne commented:
@ranisha2 would you mind updating the Sql change log with a snippet about this change? Something along the lines of

```
* Fixed issue https://github.com/Azure/azure-powershell/issues/4974
    - Removed invalid AUDIT_CHANGED_GROUP value from auditing cmdlets
```

cormacpayne commented:
On-demand: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/201/

ranisha2 commented:
@markcowl @cormacpayne 
I added a warning message when using this action group, and removed it from the list when provided.
If it's provided alone, the default value will take effect (As non providing action groups). If not, it will be simply removed (There's no mapping for default value as the user can provide a list of action groups)

ranisha2 commented on src/ResourceManager/Sql/ChangeLog.md:
```diff
@@ -19,6 +19,8 @@
 ## Current Release
 * Added ability to rename database using Set-AzureRmSqlDatabase
+* Fixed issue https://github.com/Azure/azure-powershell/issues/4974
+	- Removed invalid AUDIT_CHANGED_GROUP value from auditing cmdlets
```
@cormacpayne, @markcowl  Do I still need to include this?

cormacpayne replied:
@ranisha2 I would change the second line to say "Providing invalid AUDIT_CHANGED_GROUP value for auditing cmdlets no longer throws an error and will be removed in an upcoming release."

ranisha2 replied:
@cormacpayne Done

cormacpayne commented on src/ResourceManager/Sql/Commands.Sql/Auditing/Cmdlet/AuditingSettings/SqlDatabaseAuditingCmdletBase.cs:
```diff
@@ -54,6 +56,15 @@ protected override SqlAuditAdapter InitModelAdapter(IAzureSubscription subscript
         /// <param name="model">The model object with the data to be sent to the REST endpoints</param>
         protected override DatabaseBlobAuditingSettingsModel PersistChanges(DatabaseBlobAuditingSettingsModel model)
         {
+            if (Array.IndexOf(model.AuditActionGroup, AuditActionGroups.AUDIT_CHANGE_GROUP) > -1)
```
@ranisha2 is `model.AuditActionGroup` always going to be an initialized array?

ranisha2 replied:
@cormacpayne We perform a GetPolicy request in order to override only the provided parameters.
When not provided, we'll use the saved AuditActionGroup value, which is always initialized.

cormacpayne commented on src/ResourceManager/Sql/Commands.Sql/Auditing/Cmdlet/AuditingSettings/SqlServerAuditingCmdletBase.cs:
```diff
@@ -61,6 +64,15 @@ protected override SqlAuditAdapter InitModelAdapter(IAzureSubscription subscript
         /// <param name="baseModel">The model object with the data to be sent to the REST endpoints</param>
         protected override ServerBlobAuditingSettingsModel PersistChanges(ServerBlobAuditingSettingsModel baseModel)
         {
+            if (Array.IndexOf(baseModel.AuditActionGroup, AuditActionGroups.AUDIT_CHANGE_GROUP) > -1)
```
@ranisha2 same comment here

ranisha2 replied:
same here

cormacpayne commented:
On-demand: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/236/

markcowl commented:
second on demand run: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/238/
